doc/developer/design/20210707_qgm_sql_high_level_representation.md

```diff
@@ -0,0 +1,726 @@
+
+<!--
+// Anything currently unanswered that needs specific focus. This section may be expanded during the doc meeting as
+// other unknowns are pointed out.
+// These questions may be technical, product, or anything in-between.
+-->
+
+### Appendix A: Examples
+
+This section includes examples of how some queries look like in QGM. This visual representation will be generated
+from the representation described in the previous section. Note that having a visual way of representing the query
+is very helpful during query transformation development/troubleshooting.
+
+In this visual representation, predicates referencing columns from 1 or 2 quantifiers are represented as edges
+connecting the quantifiers involved in the predicate.
+
+#### Simple `SELECT *`
+
+![Simple SELECT *](qgm/simple-select-star.svg)
+
+#### Simple `GROUP BY`
+
+![Simple GROUP BY](qgm/simple-group-by.svg)
+
+
+#### `GROUP BY + HAVING`
+
+![GROUP BY + HAVING](qgm/simple-group-by-having.svg)
+
+Note that the having filter is just a regular predicate on the `Select` box ranging over the `Grouping` box.
+
+#### `GROUP BY` vs. `DISTINCT`
+
+As mentioned above, `GROUP BY` and `DISTINCT` use different representation in QGM. The following diagram
+shows a simple `DISTINCT` query.
+
+![DISTINCT](qgm/simple-distinct.svg)
+
+As shown below, the equivalent `GROUP BY` query is automatically converted into the query graph above
+during the normalization process.
+
+![GROUP BY DISTINCT before normalization](qgm/group-by-distinct-before.svg)
+
+![GROUP BY DISTINCT after normalization](qgm/group-by-distinct-after.svg)
+
+#### Comma join
+
+![Simple comma join](qgm/simple-comma-join.svg)
+
+#### Inner join
+
+![Simple inner join](qgm/simple-inner-join.svg)
+
+Note that the inner join above is semantically equivalent to the comma join in the previous example. Boxes 1 and 2
+represent the binary inner joins in the query, but they can be squashed into box 0, without altering the results of
+the query. In fact, the normalization step will simplify this query leaving it exactly as the one in the example above:
+
+![Simple inner join after normalization](qgm/simple-inner-join-after-normalization.svg)
+
+#### Outer join
+
+![Simple left join](qgm/simple-left-join.svg)
+
+![Simple right join](qgm/simple-right-join.svg)
+
+Note that in QGM there is no join direction, so left and right joins have the same exact representation. Only the type
+of the quantifiers change its order.
+
+Representing outer joins as explicit boxes goes against the goal of removing any sense of direction or join order
+from the query graph. For example, the following two queries are semantically equivalent but they lead to different
+query models where the join ordering is implied. We could add a normalization rule that always pushes inner joins
+through the preserving side of an outer join, assuming that it is always better to do inner joins first.
+
+![join](qgm/outer-inner-join.svg)
+![join](qgm/inner-outer-join.svg)
+
```

Comment:
There would be an outer join box with a single quantifier ranging over t3, with two correlated predicates `t1.f2 = t3.f1` and `t2.f2 = t3.f2`. There would be a top-level select box with 3 quantifiers ranging over t1, t2 and the outer join box, with a single predicate `t1.f1 = t2.f1`.

The lowering code will treat it as any other derived relation, that might be correlated, but will add the operators to add the null row if the input plan didn't return any row.

When lowering a join, we will have to build a dependency graph among the operands in the join based on correlation information, so that we can built the minimal sub-join for feeding the correlated sub-plan, and then join it with the rest of the operands in the join operator representing the join being lowered.